File: backend-vps/scraper/anichin-scraper.py
#!/usr/bin/env python3
"""
Anichin.cafe Scraper for playall.me
Scrapes donghua series + episodes + embed URLs
"""

import cloudscraper
from bs4 import BeautifulSoup
import mysql.connector
import time
import re
import sys
from datetime import datetime

# ── DB Config ─────────────────────────────────────────────────
DB = dict(host='localhost', user='streamify_user',
          password='rimbamobile2', database='streamify_db')  # isi password

BASE = 'https://anichin.cafe'

scraper = cloudscraper.create_scraper(
    browser={'browser': 'chrome', 'platform': 'windows'},
    delay=3
)
scraper.headers.update({'Accept-Language': 'id-ID,id;q=0.9,en;q=0.8'})

def db_connect():
    return mysql.connector.connect(**DB)

def slugify(text):
    text = text.lower().strip()
    text = re.sub(r'[^\w\s-]', '', text)
    return re.sub(r'[\s_-]+', '-', text)

def get_all_series():
    """Ambil semua slug seri dari /ongoing/ dan /completed/"""
    slugs = set()
    for status in ['ongoing', 'completed']:
        page = 1
        while True:
            url = f"{BASE}/{status}/page/{page}/" if page > 1 else f"{BASE}/{status}/"
            try:
                r = scraper.get(url, timeout=30)
                if r.status_code == 404:
                    break
                soup = BeautifulSoup(r.text, 'html.parser')
                items = soup.find_all('article')
                if not items:
                    break
                for it in items:
                    a = it.find('a', href=True)
                    if a and '/seri/' in a['href']:
                        slugs.add((a['href'], status))
                nxt = soup.select_one('a.next, a[rel=next], .next a')
                if not nxt:
                    break
                page += 1
                print(f"  [{status}] page {page-1} → {len(items)} series")
                time.sleep(1)
            except Exception as e:
                print(f"  Error page {page}: {e}")
                break
    return list(slugs)

def scrape_series(url, status):
    """Scrape detail seri + semua episode"""
    try:
        r = scraper.get(url, timeout=30)
        soup = BeautifulSoup(r.text, 'html.parser')

        # Title
        title = soup.find('h1')
        title = title.text.strip() if title else ''

        # Slug
        slug = url.rstrip('/').split('/seri/')[-1]

        # Poster
        img = soup.find('img', class_=lambda c: c and 'ts-post-image' in (c or ''))
        if not img:
            img = soup.select_one('.thumbook img, .thumb img')
        poster = ''
        if img:
            poster = img.get('data-src') or img.get('src', '')

        # Description
        desc_el = soup.select_one('.entry-content, .synp, .desc, [itemprop=description]')
        desc = desc_el.text.strip() if desc_el else ''

        # Genre
        genres = [a.text.strip() for a in soup.select('.genxed a, .sgenre a, [rel=tag]')]

        # Rating
        rating_el = soup.select_one('.num, .rating, [itemprop=ratingValue]')
        rating = 0.0
        if rating_el:
            try:
                rating = float(re.search(r'[\d.]+', rating_el.text).group())
            except:
                pass

        # Year
        year_el = soup.select_one('.info-content .spe span')
        year = None
        for span in soup.select('.spe span, .info-content span'):
            if re.search(r'20\d\d', span.text):
                m = re.search(r'(20\d\d)', span.text)
                if m:
                    year = int(m.group(1))
                    break

        # Episodes
        eps_links = list(dict.fromkeys([
            a['href'] for a in soup.find_all('a', href=True)
            if 'episode' in a['href'] and 'anichin.cafe' in a['href']
        ]))

        return {
            'title': title, 'slug': slug, 'poster': poster,
            'description': desc, 'genres': genres, 'rating': rating,
            'year': year, 'status': status, 'episode_urls': eps_links
        }
    except Exception as e:
        print(f"  Error scrape series {url}: {e}")
        return None

def scrape_episode_embed(ep_url):
    """Ambil embed URL dari halaman episode"""
    try:
        r = scraper.get(ep_url, timeout=30)
        soup = BeautifulSoup(r.text, 'html.parser')
        iframe = soup.find('iframe', src=True) or soup.find('iframe', attrs={'data-src': True})
        if iframe:
            return iframe.get('src') or iframe.get('data-src', '')

        # Fallback: cari di script
        for script in soup.find_all('script'):
            m = re.search(r'["\']([^"\']*anichin\.stream[^"\']*)["\']', script.text or '')
            if m:
                return m.group(1)
        return ''
    except Exception as e:
        print(f"  Error embed {ep_url}: {e}")
        return ''

def save_series(conn, series):
    """Insert/update series ke database"""
    cur = conn.cursor()
    # Cek existing
    cur.execute("SELECT id FROM content WHERE slug=%s", (series['slug'],))
    row = cur.fetchone()

    genres_str = ','.join(series['genres'])

    if row:
        content_id = row[0]
        cur.execute("""UPDATE content SET title=%s, poster_url=%s, description=%s,
            genres=%s, rating=%s, year=%s, status=%s, updated_at=NOW()
            WHERE id=%s""",
            (series['title'], series['poster'], series['description'],
             genres_str, series['rating'], series['year'],
             series['status'], content_id))
    else:
        cur.execute("""INSERT INTO content
            (title, slug, type, poster_url, description, genres, rating, year, status, created_at, updated_at)
            VALUES (%s,%s,'donghua',%s,%s,%s,%s,%s,%s,NOW(),NOW())""",
            (series['title'], series['slug'], series['poster'],
             series['description'], genres_str, series['rating'],
             series['year'], series['status']))
        content_id = cur.lastrowid

    conn.commit()
    cur.close()
    return content_id

def save_episode(conn, content_id, ep_num, embed_url, ep_url):
    """Insert/update episode"""
    cur = conn.cursor()
    cur.execute("SELECT id FROM episodes WHERE content_id=%s AND episode_number=%s",
                (content_id, ep_num))
    row = cur.fetchone()
    if row:
        cur.execute("UPDATE episodes SET embed_url=%s, source_url=%s, updated_at=NOW() WHERE id=%s",
                    (embed_url, ep_url, row[0]))
    else:
        cur.execute("""INSERT INTO episodes
            (content_id, episode_number, title, embed_url, source_url, created_at, updated_at)
            VALUES (%s,%s,%s,%s,%s,NOW(),NOW())""",
            (content_id, ep_num, f"Episode {ep_num}", embed_url, ep_url))
    conn.commit()
    cur.close()

def extract_ep_number(url):
    """Extract nomor episode dari URL"""
    m = re.search(r'episode-(\d+(?:\.\d+)?)', url)
    return float(m.group(1)) if m else 0

# ── MAIN ──────────────────────────────────────────────────────
print("🚀 Anichin Scraper mulai...")
print(f"   {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")

conn = db_connect()

# Step 1: Ambil semua series
print("\n📋 Step 1: Ambil list semua series...")
all_series = get_all_series()
print(f"   Total: {len(all_series)} series")

# Step 2: Scrape tiap series
print("\n🎬 Step 2: Scrape detail + episodes...")
for i, (url, status) in enumerate(all_series, 1):
    print(f"\n[{i}/{len(all_series)}] {url}")
    series = scrape_series(url, status)
    if not series:
        continue

    print(f"  Title   : {series['title']}")
    print(f"  Episodes: {len(series['episode_urls'])}")

    content_id = save_series(conn, series)

    # Scrape embed tiap episode
    for ep_url in series['episode_urls']:
        ep_num = extract_ep_number(ep_url)
        if ep_num == 0:
            continue
        embed = scrape_episode_embed(ep_url)
        save_episode(conn, content_id, ep_num, embed, ep_url)
        print(f"  EP {ep_num}: {embed[:60] if embed else 'no embed'}")
        time.sleep(0.5)

    time.sleep(1)

conn.close()
print("\n✅ SELESAI!")

File: backend-vps/scraper/anichin-scraper.php
<?php
/**
 * Anichin.cafe Scraper - Donghua
 */

require_once __DIR__ . '/../config/database.php';

class AnichinScraper {
    private $pdo;
    private $baseUrl = 'https://anichin.cafe';
    private $userAgent = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36';

    public function __construct() {
        $this->pdo = getDB();
    }

    private function fetch($url) {
        $ch = curl_init();
        curl_setopt_array($ch, [
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_TIMEOUT => 30,
            CURLOPT_SSL_VERIFYPEER => false,
            CURLOPT_USERAGENT => $this->userAgent,
            CURLOPT_HTTPHEADER => [
                'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
                'Accept-Language: id-ID,id;q=0.9,en-US;q=0.8',
            ],
        ]);
        $html = curl_exec($ch);
        $code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        if ($code !== 200) return null;
        return $html;
    }

    private function slugify($text) {
        $text = strtolower(trim($text));
        $text = preg_replace('/[^a-z0-9\s-]/', '', $text);
        return preg_replace('/[\s-]+/', '-', $text);
    }

    public function getAllSeries() {
        $all = [];
        foreach (['ongoing', 'completed'] as $status) {
            $page = 1;
            while (true) {
                $url = $page === 1
                    ? "{$this->baseUrl}/{$status}/"
                    : "{$this->baseUrl}/{$status}/page/{$page}/";
                $html = $this->fetch($url);
                if (!$html) break;

                preg_match_all('|href="(https://anichin\.cafe/seri/[^"]+)"|', $html, $m);
                $urls = array_unique($m[1]);
                if (empty($urls)) break;

                foreach ($urls as $u) {
                    $all[$u] = $status;
                }

                echo "  [{$status}] page {$page} → " . count($urls) . " series\n";

                if (!preg_match('|href="[^"]+/page/' . ($page+1) . '/"|', $html)) break;
                $page++;
                sleep(1);
            }
        }
        return $all;
    }

    public function scrapeSeries($url, $status) {
        $html = $this->fetch($url);
        if (!$html) return null;

        $slug = preg_replace('|.*/seri/([^/]+)/?$|', '$1', $url);

        // Title
        preg_match('|<h1[^>]*>(.*?)</h1>|si', $html, $m);
        $title = trim(strip_tags($m[1] ?? ''));

        // Poster
        preg_match('|class="[^"]*ts-post-image[^"]*"[^>]*src="([^"]+)"|', $html, $m);
        if (empty($m[1])) preg_match('|class="[^"]*thumbook[^"]*".*?<img[^>]*src="([^"]+)"|si', $html, $m);
        $poster = $m[1] ?? '';

        // Description
        preg_match('|itemprop="description"[^>]*>(.*?)</div>|si', $html, $m);
        if (empty($m[1])) preg_match('|class="[^"]*entry-content[^"]*">(.*?)</div>|si', $html, $m);
        $desc = trim(strip_tags($m[1] ?? ''));

        // Genres
        preg_match_all('|class="[^"]*genre[^"]*"[^>]*>(.*?)</a>|si', $html, $m);
        $genres = array_map('trim', $m[1] ?? []);

        // Rating
        preg_match('|itemprop="ratingValue"[^>]*>([\d.]+)|', $html, $m);
        $rating = floatval($m[1] ?? 0);

        // Year
        preg_match('|(20\d\d)|', $html, $m);
        $year = intval($m[1] ?? 0);

        // Episodes
        preg_match_all('|href="(https://anichin\.cafe/[^"]*episode[^"]+)"|', $html, $m);
        $eps = array_unique($m[1] ?? []);

        return compact('title', 'slug', 'poster', 'desc', 'genres', 'rating', 'year', 'status', 'eps');
    }

    public function scrapeEmbed($epUrl) {
        $html = $this->fetch($epUrl);
        if (!$html) return '';

        // Cari iframe src
        preg_match('|<iframe[^>]*src="([^"]*anichin\.stream[^"]*)"[^>]*>|i', $html, $m);
        if (!empty($m[1])) return $m[1];

        // Fallback iframe apapun
        preg_match('|<iframe[^>]*src="([^"]+)"[^>]*>|i', $html, $m);
        return $m[1] ?? '';
    }

    public function saveSeries($data) {
        $genres = json_encode($data['genres'] ?: []);
        $slug = $data['slug'];
        $stmt = $this->pdo->prepare("SELECT id, type FROM content WHERE slug=?");
        $stmt->execute([$slug]);
        $row = $stmt->fetch();

        if ($row && $row['type'] !== 'donghua') {
            $slug = $slug . '-donghua';
            $stmt = $this->pdo->prepare("SELECT id FROM content WHERE slug=?");
            $stmt->execute([$slug]);
            $row = $stmt->fetch();
        }

        if ($row) {
            $this->pdo->prepare("UPDATE content SET title=?,poster_url=?,description=?,genres=?,rating=?,year=?,status=? WHERE id=?")
                ->execute([$data['title'],$data['poster'],$data['desc'],$genres,$data['rating'],$data['year'],$data['status'],$row['id']]);
            return $row['id'];
        } else {
            $this->pdo->prepare("INSERT INTO content (slug,title,type,poster_url,description,genres,rating,year,status) VALUES (?,?,'donghua',?,?,?,?,?,?)")
                ->execute([$slug,$data['title'],$data['poster'],$data['desc'],$genres,$data['rating'],$data['year'],$data['status']]);
            return $this->pdo->lastInsertId();
        }
    }

    public function saveEpisode($contentId, $epNum, $embedUrl, $sourceUrl) {
        $epNum = (int)$epNum;
        $stmt = $this->pdo->prepare("SELECT id FROM episodes WHERE content_id=? AND episode_number=?");
        $stmt->execute([$contentId, $epNum]);
        $row = $stmt->fetch();

        if ($row) {
            $this->pdo->prepare("UPDATE episodes SET video_url=?,source_url=? WHERE id=?")
                ->execute([$embedUrl, $sourceUrl, $row['id']]);
        } else {
            $this->pdo->prepare("INSERT INTO episodes (content_id,episode_number,title,video_url,source_url,created_at) VALUES (?,?,?,?,?,NOW())")
                ->execute([$contentId, $epNum, "Episode {$epNum}", $embedUrl, $sourceUrl]);
        }
    }

    public function run() {
        echo "🚀 Anichin Scraper mulai...\n";
        echo "   " . date('Y-m-d H:i:s') . "\n\n";

        echo "📋 Step 1: Ambil semua series...\n";
        $allSeries = $this->getAllSeries();
        echo "   Total: " . count($allSeries) . " series\n\n";

        echo "🎬 Step 2: Scrape detail + episodes...\n";
        $i = 0;
        $total = count($allSeries);
        foreach ($allSeries as $url => $status) {
            $i++;
            echo "\n[{$i}/{$total}] {$url}\n";

            $series = $this->scrapeSeries($url, $status);
            if (!$series) { echo "  Skip!\n"; continue; }

            echo "  Title   : {$series['title']}\n";
            echo "  Episodes: " . count($series['eps']) . "\n";

            $contentId = $this->saveSeries($series);

            foreach ($series['eps'] as $epUrl) {
                preg_match('|episode-(\d+(?:\.\d+)?)|', $epUrl, $m);
                $epNum = floatval($m[1] ?? 0);
                if (!$epNum) continue;

                $embed = $this->scrapeEmbed($epUrl);
                $this->saveEpisode($contentId, $epNum, $embed, $epUrl);
                echo "  EP {$epNum}: " . substr($embed, 0, 60) . "\n";
                sleep(1);
            }
            sleep(1);
        }
        echo "\n✅ SELESAI!\n";
    }
}

$scraper = new AnichinScraper();
$scraper->run();
